### DISCUSSION TITLE
Another SD Card Initalization issue

### DISCUSSION THREAD

**Discussion by alee1234567** [polarfire-soc/polarfire-soc-documentation#149] [Bare metal embedded software] [answered]
So I tested the SD card and it is passing the single sector memory transfer test, but it is stuck at the multi sector read write test. 

![image](https://user-images.githubusercontent.com/109090085/180796586-3387e76f-4d3d-44cf-a67c-3e78e70bf05a.png)

As you can see here it is stuck at in the do while loop and never completes its transfer.

The function is multi_block_wr_rd, and the parameters are SEC_NO=1000, and USE_SDMA. 

The SD card is a 8 GB SDHC microsd.   

The SD card parameters are in the screenshot below:
![image](https://user-images.githubusercontent.com/109090085/180797350-51846973-e227-410a-9b1d-370209f743cf.png)

**Comment by padmaraob** [comment, marked as answer]
If the multi block transfer stuck in the do while loop means the MMC transfer interrupt(complete or error) not occurred

1. Can you please check the PLIC interrupts are enabled for the MMC and the MPU configuration for the MMC DMA transfer in the application (like below)?
----------------------------------------------------------
    PLIC_init();
    __enable_irq();
    PLIC_SetPriority(MMC_main_PLIC, 2u);
    PLIC_SetPriority(MMC_wakeup_PLIC, 2u);

    /* DMA init for MMC */
    loop_count = MSS_MPU_configure(MSS_MPU_MMC,
                                   MSS_MPU_PMP_REGION3,
                                   LIM_BASE_ADDRESS,
                                   LIM_SIZE,
                                   MPU_MODE_READ_ACCESS|MPU_MODE_WRITE_ACCESS|MPU_MODE_EXEC_ACCESS,
                                   MSS_MPU_AM_NAPOT,
                                   0u);
------------------------ --------------------------------

2. Apply a break point at the mmc_main_plic_IRQHandler() function in the mss-mmc.c file to check whether the MMC interrupt occurred or not .
https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/main/driver-examples/mss/mss-mmc/mpfs-emmc-sd-write-read/src/platform/drivers/mss/mss_mmc/mss_mmc.c#L2147
 
Regards
Padmarao

**Reply by alee1234567** [reply]
I did not add the statement loop_count = MSS_MPU_configure(MSS_MPU_MMC,
                               MSS_MPU_PMP_REGION3,
                               LIM_BASE_ADDRESS,
                               LIM_SIZE,
                               MPU_MODE_READ_ACCESS|MPU_MODE_WRITE_ACCESS|MPU_MODE_EXEC_ACCESS,
                               MSS_MPU_AM_NAPOT,
                               0u); 
I added it after I saw your post. 
before. So that may have been one of the issues. I then set a breakpoint at the mmc_main_plic_IRQ_Handler, and it is never hit. It just stays in the do while loop.

**Reply by alee1234567** [reply]
Okay it is fixed, I disabled the task scheduler to test another program briefly and forgot to reenable it.
